Replace filter

MarkGibbons opened the pull request:
### Description

Add the replace filter

### Issues Resolved

Adds new function.

### Contribution Check List

- [ ] All tests pass.
- [ ] New functionality includes testing.
- [ ] New functionality has been documented in the README if applicable

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sous-chefs/line/136)
<!-- Reviewable:end -->

Commit:
Add the safe option

Safe causes resources to check the lines to be inserted in the file against the regular expression used
to select lines.  If the regex matches inserted lines the resource will insert the same lines usually over
and over again making a file grow without limit. Safe can be set as a default for a filter resource or
set for each filter used in a resource.

Commit:
Add the safe option

Safe causes resources to check the lines to be inserted in the file against the regular expression used
to select lines.  If the regex matches inserted lines the resource will insert the same lines usually over
and over again making a file grow without limit. Safe can be set as a default for a filter resource or
set for each filter used in a resource.

Commit:
Add the delete_between filter

Commit:
Add the replace filter

tas50 commented on CHANGELOG.md:
```diff
@@ -1,8 +1,12 @@
 # line Cookbook CHANGELOG
 
-# Unreleased
+# Unreleased - for target version 2.3.0 in mid Janary 2018
```
```suggestion
# Unreleased - for target version 2.3.0 in mid January 2019
```

MarkGibbons replied:
Fixed

Commit:
Set the safe default for delete_between rspec testing

Commit:
Fix the filter example errors pointed out by issue #140.
Handle inserting lines specified as a string.
Update the require statements for chefspec to make it work with chefspec 3.8
Gather the common chefspec test statements into one location
Update README and documentation to include a text input example

Commit:
Fix the filter example errors pointed out by issue #140.
Handle inserting lines specified as a string.
Update the require statements for chefspec to make it work with chefspec 3.8
Gather the common chefspec test statements into one location
Update README and documentation to include a text input example

JohnRoesler commented on spec/unit/library/filter/after_spec.rb:
```diff
@@ -97,4 +98,16 @@
     expect(@filt.after(@solo_middle, [@pattern_c1, 'string1', :last])).to eq(%w(linef c1 string1 lineg))
   end
+
+  it 'should insert a line that matches the pattern by default, nil implies safe' do
```
should not?

JohnRoesler commented on spec/unit/library/filter/replace_spec.rb:
```diff
@@ -0,0 +1,75 @@
+    expect(@filt.replace(@solo_end, [@pattern_c1, @ia])).to eq(%w(linef lineg lineh line1 line2 line3))
+  end
+
+  it 'should raise error if the pattern matchs the replacement lines' do
```
```suggestion
  it 'should raise error if the pattern matches the replacement lines' do
```

JohnRoesler commented on spec/unit/library/filter/replace_spec.rb:
```diff
@@ -0,0 +1,75 @@
+    expect { @filt.replace(@current, [@pattern_c1, ['c1 match', 'c2']]) }.to raise_error(ArgumentError)
+  end
+
+  it 'should not raise error if the pattern matchs the replacement lines, force the change' do
```
```suggestion
  it 'should not raise error if the pattern matches the replacement lines, force the change' do
```

JohnRoesler reviewed (changes requested):
Some conflicts to clean up also

Commit:
Add the string to lines method to split input strings into lines

Commit:
Fix the filter_lines test

JohnRoesler committed:
Update spec/unit/library/filter/replace_spec.rb

Co-Authored-By: damacus <dan.webb@damacus.io>

JohnRoesler committed:
Update spec/unit/library/filter/replace_spec.rb

Co-Authored-By: damacus <dan.webb@damacus.io>

MarkGibbons committed:
Merge branch 'master' into replace_filter
